Issue an error on stack and queue method calls on fixed-length lists

KRVPerera opened the pull request:
## Purpose
Push/Pop functions do not natively support annotation or a way to declare that they will change the length of the list. This is not allowed in fixed-length arrays.

Fixes #18662

## Approach
`push`,`pop`, `shift` and `unshift` names are hardcoded and checked against invocations on fixed-length arrays and tuples without rest parameters to identify the invalid calls on compile time.

For unions of array types. When all the arrays are sealed types error will be given at compile time. Ignored otherwise.

## Samples
#### Code
```Ballerina
int[1] fixedLengthArray = [1];
fixedLengthArray.push(4);
```
#### Output
```
error: .::filename.bal:<line>:<column>: cannot call 'push' on fixed length list(s) of type 'int[1]'
```
## Remarks

- Some other functions like `setLength` and `removeAll` also eligible for an compile-time check but spec does not specifically request a compile-time check. 
- Spec issue # 345 may provide a way to implement this elegantly in the future.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1769,6 +1779,26 @@ private void checkInLangLib(BLangInvocation iExpr, BType varRefType) {
+            return;
+        }
+
+        if (varRefType.tag == TypeTags.ARRAY) {
```
What about unions of arrays, tuples, and combinations of arrays and tuples?

MaryamZi commented on compiler/ballerina-lang/src/main/resources/compiler.properties:
```diff
@@ -1225,12 +1225,13 @@ error.unsupported.primitive.type.reason=\
 error.invalid.attribute.reference=\
   invalid attribute reference
+error.illegal.function.array.size=\
+  cannot call ''{0}'' on fixed length array ''{1}''
```
I think we'll have to use one error and say
```
  cannot call ''{0}'' on fixed length lists
```
to handle unions.

MaryamZi replied:
Or have three errors

```
  cannot call ''{0}'' on fixed length arrays

  cannot call ''{0}'' on fixed length tuples

  cannot call ''{0}'' on fixed length lists // for unions of arrays + tuples
```

KRVPerera commented:
Need to incorporate discussion: https://github.com/ballerina-platform/ballerina-spec/issues/472

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1698,6 +1708,44 @@ private void checkInLangLib(BLangInvocation iExpr, BType varRefType) {
+    }
+
+    private void checkIllegalStorageSizeChangeMethodCall(BLangInvocation iExpr, BType varRefType) {
+        if (!modifierFunctions.contains(iExpr.name.getValue())) {
```
Shall we extract `iExpr.name.getValue()` out to a variable? We seem to be reusing it at L1747 (at times).

MaryamZi commented on langlib/langlib-test/src/test/resources/test-src/arraylib_test_negative.bal:
```diff
@@ -0,0 +1,70 @@
+}
+
+function testPushOnFixedLengthTupleUnion() {
+    [int, int][1] | [float, float][1] fixedLengthArray = <[float, float][1]> [[1.0, 2.3]];
```
```suggestion
    [int, int][1]|[float, float][1] fixedLengthArray = <[float, float][1]> [[1.0, 2.3]];
```

MaryamZi commented on langlib/langlib-test/src/test/resources/test-src/arraylib_test_negative.bal:
```diff
@@ -0,0 +1,70 @@
+
+// run time panic no compile time error
+function testPushOnFixedLengthAndDynamicTupleUnion() {
+    [int, int][1] | [float, float][] fixedLengthArray = <[float, float][1]> [[1, 2]];
```
```suggestion
    [int, int][1]|[float, float][] fixedLengthArray = <[float, float][1]> [[1, 2]];
```

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1698,6 +1708,40 @@ private void checkInLangLib(BLangInvocation iExpr, BType varRefType) {
+    private boolean isFixedLengthList(BType type) {
+        switch(type.tag) {
+            case TypeTags.ARRAY:
+                return (((BArrayType) type).state == BArrayState.CLOSED_SEALED);
```
When we reach this point the state of inferred array types is always set as `CLOSED_SEALED` right? Or we'll have to check

```suggestion
                return (((BArrayType) type).state != BArrayState.UNSEALED);
```

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -227,6 +229,14 @@ public TypeChecker(CompilerContext context) {
+    private void initModifierFunctions() {
+        this.modifierFunctions.add("push");
+        this.modifierFunctions.add("pop");
+        this.modifierFunctions.add("shift");
```
Have a small doubt regarding `shift()` on a tuple, created a spec issue - https://github.com/ballerina-platform/ballerina-spec/issues/474.

KRVPerera replied:
I think that is just a extention of this. Spec says known to be fixed length we should error out. The example given we cannot properly define right ?

MaryamZi replied:
I think we can improve this based on the spec discussion.

Can you also check the runtime behaviour?

KRVPerera replied:
added case I :  by 714edb48df99b7d411f8100bd5df60c0b2e2b05b

KRVPerera replied:
Improved for case I 714edb48df99b7d411f8100bd5df60c0b2e2b05b

MaryamZi commented on langlib/langlib-test/src/test/resources/test-src/arraylib_test_negative.bal:
```diff
@@ -0,0 +1,70 @@
+function testPushOnFixedLengthArray() {
```
Shall we add the licence header for the bal file too?

KRVPerera replied:
Test cases dont need license header right.

MaryamZi replied:
We've added the licence header in most of the new tests we've added. And given that the Java class for a test always has the licence header, I think we should include it in the bal file too.

KRVPerera replied:
added by 35e88fad66625c0dd4c8d99af1fa2ec33ec85c1d

rdhananjaya commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -227,6 +233,14 @@ public TypeChecker(CompilerContext context) {
+        initModifierFunctions();
+    }
+
+    private void initModifierFunctions() {
```
We could initialize this in a static block.

KRVPerera replied:
fixed by d9d1222ebd2f15e8eb8cc80333d36825d92b488a

rdhananjaya commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -194,6 +198,8 @@
     private BLangAnonymousModelHelper anonymousModelHelper;
     private SemanticAnalyzer semanticAnalyzer;
     private boolean nonErrorLoggingCheck = false;
+    private Set<String> modifierFunctions = new HashSet<>();
```
This could be a static variable

KRVPerera replied:
fixed by d9d1222ebd2f15e8eb8cc80333d36825d92b488a

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1695,7 +1707,60 @@ private void checkInLangLib(BLangInvocation iExpr, BType varRefType) {
+            resultType = symTable.semanticError;
+        }
+
+        if ((varRefType.tag == TypeTags.TUPLE) && hasDifferentTypeThanRest((BTupleType) varRefType) &&
```
What about unions of tuples?

This should be an error for tuple unions which will always fail too?

KRVPerera replied:
Done
https://github.com/ballerina-platform/ballerina-lang/pull/22318/commits/2fe86c1ae95def6e356be4f14ba27b2d3a8301f1

KRVPerera commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -176,9 +176,14 @@
+    private static final String FUNCTION_PUSH_NAME = "push";
+    private static final String FUNCTION_POP_NAME = "pop";
+    private static final String FUNCTION_SHIFT_NAME = "shift";
+    private static final String FUNCTION_UNSHIFT_NAME = "unshift";
```
Done
https://github.com/ballerina-platform/ballerina-lang/pull/22318/commits/ad713781a969cc96f320ad7789b45f3a697b1421

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1710,7 +1722,80 @@ private void checkInLangLib(BLangInvocation iExpr, BType varRefType) {
+
+        if (isFixedLengthList(varRefType)) {
+            dlog.error(iExpr.name.pos, DiagnosticCode.ILLEGAL_FUNCTION_CHANGE_LIST_SIZE, invocationName, varRefType);
+            resultType = symTable.semanticError;
```
We can return here right?

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1710,7 +1722,80 @@ private void checkInLangLib(BLangInvocation iExpr, BType varRefType) {
+            resultType = symTable.semanticError;
+        }
+
+        if ((varRefType.tag == TypeTags.TUPLE) && hasDifferentTypeThanRest((BTupleType) varRefType) &&
+                (invocationName.compareTo(FUNCTION_NAME_SHIFT) == 0)) {
+            dlog.error(iExpr.name.pos, DiagnosticCode.ILLEGAL_FUNCTION_CHANGE_TUPLE_SHAPE, invocationName,
+                       varRefType);
+            resultType = symTable.semanticError;
+        }
+
+        if ((varRefType.tag == TypeTags.UNION) && (invocationName.compareTo(FUNCTION_NAME_SHIFT) == 0)) {
+            BUnionType unionVarRef = (BUnionType) varRefType;
+            boolean allTuplesHasFixedShape = true;
+            for (BType member : unionVarRef.getMemberTypes()) {
+                if (member.tag != TypeTags.TUPLE) {
+                    break;
+                }
+                if (!hasDifferentTypeThanRest((BTupleType) varRefType)) {
+                    allTuplesHasFixedShape = false;
+                    break;
+                }
+            }
+            if (allTuplesHasFixedShape) {
+                dlog.error(iExpr.name.pos, DiagnosticCode.ILLEGAL_FUNCTION_CHANGE_TUPLE_SHAPE, invocationName,
+                           varRefType);
+                resultType = symTable.semanticError;
+            }
+        }
```
Shall we extract this tuple check logic to a function like `isShiftOnIncompatibleTuples` and handle both individual tuples and unions?

Then we can do 

```java
if (isShiftOnIncompatibleTuples(varRefType, invocationName)) {
      dlog.error(iExpr.name.pos, DiagnosticCode.ILLEGAL_FUNCTION_CHANGE_TUPLE_SHAPE, invocationName,
                 varRefType);
      resultType = symTable.semanticError;
}
```

MaryamZi commented on compiler/ballerina-lang/src/main/resources/compiler.properties:
```diff
@@ -1243,3 +1243,9 @@ error.deprecated.parameters.documentation.not.allowed=\
+  cannot call ''{0}'' on fixed length list(s) of type ''{1}''
+
+error.illegal.function.change.tuple.shape=\
+  cannot call ''{0}'' on tuple(s) of type ''{1}''; cannot violate inherent type
```
```suggestion
  cannot call ''{0}'' on tuple(s) of type ''{1}'': cannot violate inherent type
```

MaryamZi commented:
Do we have positive/negative runtime tests for modification attempts on unions with combinations of arrays, tuples, and arrays and tuples, where at least one member has a fixed length?
e.g.,
```ballerina
int[4]|int[]

[string, string...]|[string, int]


int[4]|[string...] // also is the resultant type only `string` here? 
int[]|[string, boolean]
```

And also tests for `shift()` on non-fixed length tuple that may or may not succeed.

e.g.,
```ballerina
public function main() {
    // Case I
    [int, int...] i = [0, 1];
    int x = i.shift();

    // Case II
    [int, int...] i = [0];
    int x = i.shift();
}
```

KRVPerera commented:
I have added test cases and created issues.

KRVPerera commented on langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal:
```diff
@@ -513,6 +513,25 @@ function testInvalidPushOnUnionOfSameBasicType() {
+    testShiftOnTupleWithoutValuesForRestParameter();
+}
+
+function testShiftOnTupleWithoutValuesForRestParameter() {
```
may change the error message in the future.

KRVPerera commented:
I have added a few test cases by https://github.com/ballerina-platform/ballerina-lang/pull/22318/commits/c2cb9ab44b3c92610eb1049865be003fec253a28. I also fixed the logic by https://github.com/ballerina-platform/ballerina-lang/pull/22318/commits/3daa2b17c60d926b53c07da535dd7bed59c16228. Created issues for existing problems.
https://github.com/ballerina-platform/ballerina-lang/issues/22771 : Union of same basic types should resolve to valid lang.lib method at compile time
    This blocks 

```Ballerina
int[4]|[string...] // also is the resultant type only `string` here?
int[]|[string, boolean]
```

https://github.com/ballerina-platform/ballerina-lang/issues/22765: Bad Sad Error for shift operation on same type union of lists
    This blocks
```Ballerina
int[4]|int[]
```

https://github.com/ballerina-platform/ballerina-lang/issues/22769: Tuple with same basic type and rest parameter should allow `shift` in some scenarios
This issue blocks case I and case II.

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:
```diff
@@ -1821,7 +1833,86 @@ private void checkInLangLib(BLangInvocation iExpr, BType varRefType) {
+            BUnionType unionVarRef = (BUnionType) varRefType;
+            boolean allMemberAreFixedShapeTuples = true;
+            for (BType member : unionVarRef.getMemberTypes()) {
+                if (member.tag != TypeTags.TUPLE) {
```
When we are fixing https://github.com/ballerina-platform/ballerina-lang/issues/22771, I think we should refactor `checkIllegalStorageSizeChangeMethodCall` and this method a bit.

At L1873 if the `invocationName` is `shift` we can call something like

```java
private boolean isShiftValidOnLists(BType varRefType) {
     int tag = varRefType.tag;

     if (tag == TypeTags.ARRAY) {
         return !isFixedLengthList(varRefType);
     }

     if (tag == TypeTags.TUPLE) {
         return !hasDifferentTypeThanRest((BTupleType) member);
     }

     // Union type here.
    BUnionType unionVarRef = (BUnionType) varRefType;
    for (BType member : unionVarRef.getMemberTypes()) {
        if (!isShiftValidOnLists(member)) {
            return false;
        }
    }
    return true;
}
```
